src/thread_task.rs

```diff
@@ -813,869 +813,746 @@ unsafe fn delayed_fg_task(c: *const Rav1dContext, ttd: *mut TaskThreadData) {
+                let first: c_uint = ::core::intrinsics::atomic_load_seqcst(&mut (*ttd).first);
+                f = &mut *((*c).fc).offset(first.wrapping_add(i).wrapping_rem((*c).n_fc) as isize)
+                    as *mut Rav1dFrameContext;
+                if ::core::intrinsics::atomic_load_seqcst(
+                    &mut (*f).task_thread.init_done as *mut atomic_int,
+                ) != 0
+                {
+                    continue;
+                }
+                t = (*f).task_thread.task_head;
+                if t.is_null() {
+                    continue;
+                }
+                if (*t).type_0 as c_uint == RAV1D_TASK_TYPE_INIT as c_int as c_uint {
+                    found = true;
+                    break;
+                }
+                if (*t).type_0 as c_uint == RAV1D_TASK_TYPE_INIT_CDF as c_int as c_uint {
+                    // XXX This can be a simple else, if adding tasks of both
+                    // passes at once (in dav1d_task_create_tile_sbrow).
+                    // Adding the tasks to the pending Q can result in a
+                    // thread merging them before setting init_done.
+                    // We will need to set init_done before adding to the
+                    // pending Q, so maybe return the tasks, set init_done,
+                    // and add to pending Q only then.
+                    let p1 = (if !((*f).in_cdf.progress).is_null() {
+                        ::core::intrinsics::atomic_load_seqcst((*f).in_cdf.progress)
+                    } else {
+                        1 as c_int as c_uint
+                    }) as c_int;
+                    if p1 != 0 {
+                        ::core::intrinsics::atomic_or_seqcst(
+                            &mut (*f).task_thread.error,
+                            (p1 == TILE_ERROR) as c_int,
+                        );
+                        found = true;
+                        break;
+                    }
+                }
+            }
+        }
+        if !found {
+            'decoding: while (*ttd).cur < (*c).n_fc {
+                let first_0: c_uint = ::core::intrinsics::atomic_load_seqcst(&mut (*ttd).first);
+                f = &mut *((*c).fc)
+                    .offset(first_0.wrapping_add((*ttd).cur).wrapping_rem((*c).n_fc) as isize)
+                    as *mut Rav1dFrameContext;
+                merge_pending_frame(f);
+                prev_t = (*f).task_thread.task_cur_prev;
+                t = if !prev_t.is_null() {
+                    (*prev_t).next
+                } else {
+                    (*f).task_thread.task_head
+                };
+                while !t.is_null() {
+                    if (*t).type_0 as c_uint == RAV1D_TASK_TYPE_INIT_CDF as c_int as c_uint {
+                        // goto next;
+                        prev_t = t;
+                        t = (*t).next;
+                        (*f).task_thread.task_cur_prev = prev_t;
+                        continue;
```

Comment:
I think it'd be helpful to have every `continue` and `break` labeled in this `fn` due to the sheer complexity.